### PR TITLE
#11793 Delete User Dependency before delete the User

### PR DESCRIPTION
**Pull request by BodorAnnamaria**
https://github.com/metasfresh/metasfresh/issues/11793

### CODE REVIEW

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java [hunk 1]
```diff
@@ -318,4 +326,75 @@ public void save(@NonNull final org.compiere.model.I_AD_User user)
 		return OrgMappingId.optionalOfRepoId(sourceUserRecord.getAD_Org_Mapping_ID());
 	}
 
+	@Override
+	public void deleteUserPreferenceByUserId(final UserId userId)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java:330]: move it to ValuePreferenceDAO

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java [hunk 2]
```diff
@@ -318,4 +326,75 @@ public void save(@NonNull final org.compiere.model.I_AD_User user)
...
+	{
+		queryBL.createQueryBuilder(I_AD_Preference.class)
+				.addEqualsFilter(I_AD_Preference.COLUMNNAME_AD_User_ID, userId)
+				.create()
+				.delete();
+	}
+
+	@Override
+	public void deleteUserOrgAccessByUserId(final UserId userId)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java:339]: move it to UserRolePermissionsDAO

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java [hunk 3]
```diff
@@ -318,4 +326,75 @@ public void save(@NonNull final org.compiere.model.I_AD_User user)
...
+		queryBL.createQueryBuilder(I_C_OrgAssignment.class)
+				.addEqualsFilter(I_C_OrgAssignment.COLUMNNAME_AD_User_ID, userId)
+				.create()
+				.delete();
+	}
+
+	@Override
+	public void deletePInstancesForUserId(final UserId userId)
+	{
+		queryBL.createQueryBuilder(I_AD_PInstance.class)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java:395]: change the AD_PInstance.AD_User_ID:
* drop the FK
* in Table & Columns, for AD_user_ID column, flag it as no FK

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java (hunk 1 above)
Comment (BodorAnnamaria) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java]: done

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java (hunk 2 above)
Comment (BodorAnnamaria) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java]: Done

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java (hunk 3 above)
Comment (BodorAnnamaria) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java]: FK droped

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java [hunk 4]
```diff
@@ -18,7 +18,15 @@
...
 import org.adempiere.util.proxy.Cached;
+import org.compiere.model.I_AD_PInstance;
+import org.compiere.model.I_AD_Preference;
 import org.compiere.model.I_AD_User;
+import org.compiere.model.I_AD_UserMail;
+import org.compiere.model.I_AD_UserQuery;
+import org.compiere.model.I_AD_User_OrgAccess;
+import org.compiere.model.I_AD_User_Roles;
+import org.compiere.model.I_AD_User_Substitute;
+import org.compiere.model.I_C_OrgAssignment;
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java:29]: pls Organize Imports. I think some of them shall vanish (e.g. I_AD_PInstance, I_AD_Preference etc)

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java [hunk 5]
```diff
@@ -318,4 +326,48 @@ public void save(@NonNull final org.compiere.model.I_AD_User user)
...
+	public void deleteUserQueryByUserId(final UserId userId)
+	{
+		queryBL.createQueryBuilder(I_AD_UserQuery.class)
+				.addEqualsFilter(I_AD_UserQuery.COLUMNNAME_AD_User_ID, userId)
+				.create()
+				.delete();
+	}
+
+	@Override
+	public void deleteUserRolesByUserId(final UserId userId)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java:357]: pls, move it to RoleDAO.

Basically, each database table shall be handled in its own Repository and nowhere else!

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java (hunk 5 above)
Comment (BodorAnnamaria) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java]: Done

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java (hunk 4 above)
Comment (BodorAnnamaria) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/user/api/impl/UserDAO.java]: cleaned

**Pull request merged**